cpp/oneapi/dal/algo/kmeans_init/backend/gpu/compute_kernel_distr_plus_plus_dense_dpc.cpp

```diff
@@ -0,0 +1,763 @@
+                         pr::ndview<Float, 1>& values,
+                         pr::ndview<Float, 1>& bounds,
+                         pr::ndview<Index, 1>& indices,
+                         const bk::event_vector& deps) {
+    const auto rank = comm.get_rank();
+    const auto rank_count = comm.get_rank_count();
+
+    ONEDAL_ASSERT(bounds.has_mutable_data());
+    ONEDAL_ASSERT((rank_count + 1) == bounds.get_count());
+
+    ONEDAL_ASSERT(values.has_mutable_data());
+    const auto local_count = values.get_count();
+
+    // Computes local cumulative sum
+    //auto fill_bnds_event = pr::fill(queue, bounds, Float(0), deps);
+    auto last_event = pr::cumulative_sum_1d(queue, values, deps);
+    auto fill_event = pr::fill(queue, bounds, Float(0), deps);
+
+    const Float* const last_val_ptr = values.get_data() + local_count - 1;
+
+    if (rank_count > 1) {
+        ONEDAL_PROFILER_TASK(find_indices.boundaries, queue);
+
+        auto bnds_arr = array<Float>::wrap(queue, bounds.get_mutable_data() + 1, rank_count);
+        const Float* const last_val_ptr = values.get_data() + local_count - 1;
+        auto last_val_arr = array<Float>::wrap(queue, last_val_ptr, 1);
+
+        sycl::event::wait_and_throw({ last_event, fill_event });
+        comm.allgather(last_val_arr, bnds_arr).wait();
+
+        auto cumsum_event = pr::cumulative_sum_1d(queue, bounds);
+
+        const auto adjustment = bounds.get_slice(rank, rank + 1);
+        last_event = add_number(queue, adjustment, values, { cumsum_event });
+    }
+    else {
+        auto slice = bounds.get_slice(rank + 1, rank + 2);
+        const auto set = [=](auto&, auto* ptr) -> Float {
+            return *ptr;
+        };
+        last_event =
+            element_wise(queue, set, slice, last_val_ptr, slice, { last_event, fill_event });
+    }
+
+    constexpr auto alignment = pr::search_alignment::left;
+    auto search_event = pr::search_sorted_1d(queue,
+                                             alignment, //
+                                             values,
+                                             points,
+                                             indices,
+                                             { last_event });
+
+    auto fix_event = fix_indices(queue, comm, offset, points, bounds, indices, { search_event });
+
+    return fix_event;
+}
+
+template <typename Method, typename Task, typename Float, pr::ndorder order>
+compute_result<Task> implementation(const bk::context_gpu& ctx,
+                                    const detail::descriptor_base<Task>& params,
+                                    const pr::ndview<Float, 2, order>& samples,
+                                    const bk::event_vector& deps = {}) {
+    using dal::backend::operator+;
+
+    constexpr pr::sum<Float> sum{};
+    constexpr pr::square<Float> square{};
+
+    auto& queue = ctx.get_queue();
+    auto& comm = ctx.get_communicator();
+
+    auto rng = std::mt19937(params.get_seed());
```

Comment:
1) For now we have hardcoded RNGs all over the library. I guess it can be fixed by using some abstract entity instead but so far there is no need in such complex logic
2) In case of number of clusters $C=128$ (it is reasonably high number) by default it will result in $1 + (C - 1) \cdot (2 + \log C) \approx  872$ generated 64 bit numbers. In my mind we can neglect performance impact of the standard C++ implementation in pretty much all reasonable use cases